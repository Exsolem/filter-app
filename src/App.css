.App {
    text-align: center;
}


/* img {
    max-width: 30vmin;
    max-height: 30vh;
}

ul>li {
    display: flex;
    flex-flow: row nowrap;
    justify-content: center;
    align-items: center;
    width: 90vw;
    margin: 5vmin;
    border-bottom: 1px solid black;
}

li>span {
    width: 20vw;
    font-size: 3vmin;
} */