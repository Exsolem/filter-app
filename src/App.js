import React from 'react'
import { BrowserRouter as Router, Route } from 'react-router-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { mapStateToProps, mapDispatchToProps } from './ducks/products'
import { AppContainer, AppDiv, AppImg, AppMenu } from "./styled/styledComponents"
import Products from './products/products'
import CategoriesList from "./categoriesList/categoriesList";


export class App extends React.Component {
    componentDidMount() {
        this.props.fetchProducts();
    }

    componentDidUpdate(prevProps, prevState) {
        if (this.props.currentCategory && prevProps.products.length !== this.props.products.length) {
            this.props.getFilteredProducts();
            this.props.getSearchedProducts();
        }
        if (this.props.search && prevProps.search.length !== this.props.search.length) {
            this.props.getSearchedProducts();
        }
    }

    render() {
        return (
            <AppContainer className="col-lg-12">
                <Router>
                    <AppDiv>
                        <AppMenu>
                            <CategoriesList
                                currentCategory={this.props.currentCategory}
                                categories={this.props.categories}
                                search={this.props.search}
                                getSearchValue={this.props.getSearchValue}
                                getCurrentCategory={this.props.getCurrentCategory}
                                getFilteredProducts={this.props.getFilteredProducts}
                                getSearchedProducts={this.props.getSearchedProducts}
                            />
                        </AppMenu>
                        {this.props.isLoading ? (
                            <AppImg
                                src="https://i.pinimg.com/originals/3f/2c/97/3f2c979b214d06e9caab8ba8326864f3.gif"
                                alt="loader"
                            />
                        ) : this.props.errorOccurred ? (
                            <p>Error, try again</p>
                        ) : (
                            <Route path='/' render={props => (
                                <Products
                                    {...props}
                                    getFilteredProducts={this.props.getFilteredProducts}
                                    filteredProducts={this.props.filteredProducts}
                                    searchedProducts={this.props.searchedProducts}
                                    search={this.props.search}
                                    products={this.props.products}
                                    getSearchValue={this.props.getSearchValue}
                                    getSearchedProducts={this.props.getSearchedProducts}
                                    getCurrentCategory={this.props.getCurrentCategory}
                                />
                            )}/>
                        )}
                    </AppDiv>
                </Router>
            </AppContainer>
        );
    }

}

App.propTypes = {
    currentCategory: PropTypes.string,
    category: PropTypes.string,
    getFilteredProducts: PropTypes.func,
    fetchProducts: PropTypes.func,
    filteredProducts: PropTypes.array,
    searchedProducts: PropTypes.array,
    search: PropTypes.string,
    products: PropTypes.array,
    getSearchValue: PropTypes.func,
    getSearchedProducts: PropTypes.func,
    getCurrentCategory: PropTypes.func,
    Products: PropTypes.element,
    Category: PropTypes.element
};
export default connect(mapStateToProps, mapDispatchToProps)(App);
