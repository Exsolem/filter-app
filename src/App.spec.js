import React from 'react'
import { shallow } from 'enzyme'

import  { App } from './App'

describe('App', () => { // групируем с помощью describe все тесты для контейнера News
    const mockFetchProducts = jest.fn() // создали mock-функцию
    const props = {
        isLoading: false,
        loadSuccess: false,
        errorOccurred: false,
        products: [],
        filteredProducts: [],
        searchedProducts: [],
        categories: [],
        search: '',
        currentCategory: '',
        fetchProducts: mockFetchProducts
    };

    describe('App initial', () => {
        const app = shallow(<App {...props}/>);
        it('renders properly', () => {
            expect(app.debug()).toMatchSnapshot()
        });
        it('dispatches the `fetchProducts()` method it receives from props', () => {
            expect(mockFetchProducts).toHaveBeenCalled() // создали ожидание с нужным ассертом
        });
        it('not rendered products', () => {
            expect(app.find('Products')).toHaveLength(0);
        });
        it('not rendered CategoriesList',() => {
            expect(app.find('CategoriesList')).toHaveLength(0);
        })
    });
    describe('Products is loading', () => {
        const nextProps = {
            ...props,
            isLoading: true
        };
        const app = shallow(<App {...nextProps} />);
        it('renders properly', () => {
            expect(app.debug()).toMatchSnapshot()
        });
        it('render loader', () => {
            expect(app.find('AppImg')).toHaveLength(1);
        });
        it('not render Products', () => {
            expect(app.find('Products')).toHaveLength(0)
        });
        it('not render error', () => {
            expect(app.find('p')).toHaveLength(0)
        })
    })
    describe('Products is loaded with error', () => {
        const nextProps = {
            ...props,
            errorOccurred: true
        };
        const app = shallow(<App {...nextProps} />);
        it('renders properly', () => {
            expect(app.debug()).toMatchSnapshot()
        });
        it('renders error', () => {
            expect(app.find('p').text()).toEqual('Error, try again');
        });
        it('not render loader', () => {
            expect(app.find('AppImg')).toHaveLength(0);
        });
        it('not render Products', () => {
            expect(app.find('Products')).toHaveLength(0)
        });

    });
    describe('Products loading is success', () => {
        const nextProps = {
            ...props,
            loadSuccess: true
        };
        const app = shallow(<App {...nextProps} />);
        it('renders properly', () => {
            expect(app.debug()).toMatchSnapshot()
        });
        it('rendered Products', () => {
            expect(app.find('Products')).toHaveLength(0)
        });
        it('not render loader', () => {
            expect(app.find('AppImg')).toHaveLength(0);
        });
        it('not render error', () => {
            expect(app.find('p')).toHaveLength(0)
        })
    })
});