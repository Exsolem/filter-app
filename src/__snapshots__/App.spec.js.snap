// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App initial renders properly 1`] = `
"<AppContainer className=\\"col-lg-12\\">
  <BrowserRouter>
    <AppDiv>
      <AppMenu>
        <withRouter(CategoriesList) currentCategory=\\"\\" categories={{...}} search=\\"\\" getSearchValue={[undefined]} getCurrentCategory={[undefined]} getFilteredProducts={[undefined]} getSearchedProducts={[undefined]} />
      </AppMenu>
      <Route path=\\"/\\" render={[Function: render]} />
    </AppDiv>
  </BrowserRouter>
</AppContainer>"
`;

exports[`App Products is loaded with error renders properly 1`] = `
"<AppContainer className=\\"col-lg-12\\">
  <BrowserRouter>
    <AppDiv>
      <AppMenu>
        <withRouter(CategoriesList) currentCategory=\\"\\" categories={{...}} search=\\"\\" getSearchValue={[undefined]} getCurrentCategory={[undefined]} getFilteredProducts={[undefined]} getSearchedProducts={[undefined]} />
      </AppMenu>
      <p>
        Error, try again
      </p>
    </AppDiv>
  </BrowserRouter>
</AppContainer>"
`;

exports[`App Products is loading renders properly 1`] = `
"<AppContainer className=\\"col-lg-12\\">
  <BrowserRouter>
    <AppDiv>
      <AppMenu>
        <withRouter(CategoriesList) currentCategory=\\"\\" categories={{...}} search=\\"\\" getSearchValue={[undefined]} getCurrentCategory={[undefined]} getFilteredProducts={[undefined]} getSearchedProducts={[undefined]} />
      </AppMenu>
      <AppImg src=\\"https://i.pinimg.com/originals/3f/2c/97/3f2c979b214d06e9caab8ba8326864f3.gif\\" alt=\\"loader\\" />
    </AppDiv>
  </BrowserRouter>
</AppContainer>"
`;

exports[`App Products loading is success renders properly 1`] = `
"<AppContainer className=\\"col-lg-12\\">
  <BrowserRouter>
    <AppDiv>
      <AppMenu>
        <withRouter(CategoriesList) currentCategory=\\"\\" categories={{...}} search=\\"\\" getSearchValue={[undefined]} getCurrentCategory={[undefined]} getFilteredProducts={[undefined]} getSearchedProducts={[undefined]} />
      </AppMenu>
      <Route path=\\"/\\" render={[Function: render]} />
    </AppDiv>
  </BrowserRouter>
</AppContainer>"
`;
